### PR TITLE
Make SSL_read return the old behaviour and document it.

### PR DESCRIPTION
**Pull request by kroeckx**
This reverts commit 4880672a9b41a09a0984b55e219f02a2de7ab75e.

Fixes: #1903

### CODE REVIEW

**Comment by mattcaswell** [comment]
I think we should fix the SSL_get_error() documentation where it talks about EOF too.

**Comment by kroeckx** [comment]
I'm not sure what to write there. I assume it's about SSL_ERROR_SYSCALL. I assume that there was actually a system call that failed in that case. But I have no idea if we have any BIO that returns the unexpected value in that case.

**Comment by kroeckx** [comment]
For SSL_read_ex we now say to call SSL_get_error(), I guess that's also good advice for SSL_read(), instead of looking at the 0 / -1?

I'm also unsure how to actually know if something is retryable with SSL_read_ex() since it can't return -1.

**Comment by mattcaswell** [comment]
> I'm not sure what to write there. I assume it's about SSL_ERROR_SYSCALL. I assume that there was actually a system call that failed in that case. But I have no idea if we have any BIO that returns the unexpected value in that case.

Yes its for SSL_ERROR_SYSCALL. The whole discussion on examining the value of `ret` is just wrong IMO. It may be true for some BIOs and in some situations, but it is not a reliable indicator. I would just say that "Some non-recoverable error occurred in a system call.", or something like that.

> For SSL_read_ex we now say to call SSL_get_error(), I guess that's also good advice for SSL_read(), instead of looking at the 0 / -1?

Yes, I see a 0 / -1 return from SSL_read() as the same thing. You can't read anything into either value.

> I'm also unsure how to actually know if something is retryable with SSL_read_ex() since it can't return -1.

Call SSL_get_error(). If it returns SSL_ERROR_WANT_* then it's retryable. SSL_get_error() tests the retryability of the underlying BIO, i.e. in SSL_get_error():

````
    if (SSL_want_read(s)) {
        bio = SSL_get_rbio(s);
        if (BIO_should_read(bio))
            return (SSL_ERROR_WANT_READ);
````
and
````
    if (SSL_want_write(s)) {
        /*
         * Access wbio directly - in order to use the buffered bio if
         * present
         */
        bio = s->wbio;
        if (BIO_should_write(bio))
            return (SSL_ERROR_WANT_WRITE);
````

**Comment by mattcaswell** [comment]
BTW, travis is failing on this PR. Probably due to commit a34ac5b8b9.

**Comment by kroeckx** [comment]
So are you saying that SSL_ERROR_SYSCALL is never something that can be retried, unlike what the documentation says? It now says to look at ret, which we're here documenting as something you can't rely on.

**Comment by kroeckx** [comment]
Also, if SSL_get_error() is a good way to detect that it's retryable or not, why don't we have SSL_read() call that so it knows if it should return -1 or 0?

**Comment by kroeckx** [comment]
The current SSL_read() in master can't even return -1 anymore, since it returns whatever SSL_read_ex() returned which always return 0 on failure.

So the documentation for SSL_ERROR_SYSCALL says that 0 is EOF. But we're now saying that it wasn't always the case and might actually have been retryable. But then we changed it to always return 0 even if there wasn't an EOF. This is clearly wrong for people using the old documentation. And you're now even saying that there is no way to detect that we got an unexpected EOF and must always retry even if there was an EOF.

I'm currently not sure what to do. But it's at least clear that whatever this patch was doing is not good enough for master and we need a different patch for master an 1.1.0.

**Comment by kroeckx** [comment]
The more I look at this, the more things are inconsistent. It seem to randomly returns 0 or -1 all over the place.

**Comment by kroeckx** [comment]
I've pushed a new commit

File: doc/man3/SSL_read.pod [hunk 1]
```diff
@@ -87,22 +87,15 @@ with the same arguments.
 
 =head1 RETURN VALUES
 
-SSL_read_ex() and SSL_peek_ex() will return 1 for success or 0 for failure.
-Success means that 1 or more application data bytes have been read from the SSL
-connection. Failure means that no bytes could be read from the SSL connection.
-Failures can be retryable (e.g. we are waiting for more bytes to
-be delivered by the network) or non-retryable (e.g. a fatal network error). In
-the event of a failure call L<SSL_get_error(3)> to find out the reason which
-indicates whether the call is retryable or not.
```
Comment (mattcaswell) [inline doc/man3/SSL_read.pod]: Why are you deleting this? This is the intended semantics (although actually SSL_read_ex()/SSL_peek_ex() don't currently do this!....but that should be fixed)

File: doc/man3/SSL_get_error.pod [hunk 2]
```diff
@@ -126,6 +126,13 @@ OpenSSL error queue contains more information on the error.
 
 =back
 
+=head1 BUGS
+
+In case of SSL_ERROR_SYSCALL, B<ret> is meant to indicate that an error is
+retryable or not.
```
Comment (mattcaswell) [inline doc/man3/SSL_get_error.pod]: It is never be retryable if you get SSL_ERROR_SYSCALL. SSL_get_error() should give you an SSL_ERROR_WANT_* return if it is retryable. The previous description of SSL_ERROR_SYSCALL is just wrong I think,

**Comment by mattcaswell** [comment]
> The more I look at this, the more things are inconsistent. It seem to randomly returns 0 or -1 all over the place.

Yes, I know. But trying to make things more consistent is how we got to where we are now. :-(

> The current SSL_read() in master can't even return -1 anymore, since it returns whatever SSL_read_ex() returned which always return 0 on failure.

Actually I think its the other way around. SSL_read_ex() isn't doing what its supposed to be doing and still returns 0/-1. That should be fixed (without breaking the return value of SSL_read())

> So are you saying that SSL_ERROR_SYSCALL is always something that can be retried, unlike what the documentation says?

No. SSL_ERROR_SYSCALL can never be retried. If something is retryable you should get an SSL_ERROR_WANT_* return.

**Comment by kroeckx** [comment]
New commit pushed

**Comment by kroeckx** [comment]
So I tried #1964, but that doesn't seem to work out.

**Comment by mattcaswell** [comment]
Approved, for master - but I think we need to do this for 1.1.0, although I think this probably won't apply cleanly.

**Comment by kroeckx** [comment]
Yes, I'll make a different pull request for 1.1.0

**Comment by kroeckx** [comment]
So in #1966 (for 1.1.0) I actually get a test failure because SSL_get_error() looks at the return value and in case that was 0 it returns either SSL_ERROR_ZERO_RETURN or SSL_ERROR_SYSCALL. In the master branch it returns -1 instead and you get SSL_ERROR_WANT_WRITE.

The SSL_ERROR_SYSCALL is just the wrong return value there, and I wonder why it returned 0 in 1.1.0 and -1 in master.

So I currently don't even see a way of documenting this in a way that makes sense and still works with older versions.

**Comment by kroeckx** [comment]
So of course master returned -1, that's why we have this merge request. I was looking at the wrong branch.

So the difference in behavior between this (for master) and #1966 (for 1.1.0) is that SSL_get_error() will also return one of the WANT errors in case 0 was returned. I wonder if we should also backport that.

**Pull request closed without merging**